semgrep-core/Synthesizing/Pattern_from_Targets.ml

```diff
@@ -0,0 +1,381 @@
+(* Emma Jin
+ *
+ * Copyright (C) 2020 r2c
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License (GPL)
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * file license.txt for more details.
+*)
+open AST_generic
+open Common
+module Set = Set_
+
+exception InvalidSubstitution
+exception UnsupportedTargetType
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* The main target intersection algorithm.
+ * Helper functions are very similar to Pattern_from_Code --- refactor?
+ *
+ * related work:
+ *  - coccinelle spinfer?
+*)
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type stage = DONE | ANY of any | LN of any
+type env = { prev : any; count : int; mapping : (expr * expr) list; }
+type pattern_instrs = (env * any * ((stage * ((any -> any) -> any -> any)) list)) list
```

Comment:
this type is quite complicated. At least give an alias to the 3rd element, with a good
name so we can get an idea of what it is useful for.